### DISCUSSION TITLE
How do I do ADT in hy?

### DISCUSSION THREAD

**Discussion by glyh** [hylang/hy#2392] [Q&A] [answered]
Here's a link of how do you do [ADT](https://stackoverflow.com/questions/16258553/how-can-i-define-algebraic-data-types-in-python) in python.

I'm aware of we have match, but how would you implement a recursive tree with it?

**Comment by Kodiologist** [comment]
We've supported Python 3.11 for a while, so you can translate any of those examples to Hy in a literal fashion and the effect should be the same.

**Reply by glyh** [reply]
How do you translate this, then?
```py
Shape = Point | Circle | Rectangle
```
EDIT:

Nvm I found out:
```hy
(setv Shape (| Point Circle Rectangle))
```

**Reply by glyh** [reply]
How do you annotate with `@dataclass` then?

**Reply by Kodiologist** [reply]
It sounds like you're figuring this out while you type. Try working on this yourself for a while and then come back if you're stuck.

**Comment by glyh** [comment, marked as answer]
I figured it out, something is not well-documented, but here's the final code for [ADT example on SO](https://stackoverflow.com/questions/16258553/how-can-i-define-algebraic-data-types-in-python):
```hy
(import dataclasses [dataclass])

(defclass [dataclass] Point []
  (annotate x float)
  (annotate y float))

(defclass [dataclass] Circle []
  (annotate x float)
  (annotate y float)
  (annotate r float))

(defclass [dataclass] Rectangle []
  (annotate x float)
  (annotate y float)
  (annotate w float)
  (annotate h float))

(setv Shape (| Point Circle Rectangle))

(defn print-shape [#^Shape shape]
  (match shape
    (Point x y)    (print f"Point {x} {y}")
    (Circle x y r) (print f"Circle {x} {y} {r}")
    (Rectangle x y w h) (print f"Circle {x} {y} {w} {h}")))

(print-shape (Point 1 2))
(print-shape (Circle 3 5 7))
(print-shape (Rectangle 11 13 17 19))
(print-shape 4)
```
I would say, though, it's kinda ugly. so I may write a macro for this.
